### PR TITLE
Importer

### PR DESCRIPTION
**Pull request by sean-yeoh**
### Summary
Currently, after a user uploads a file, it doesn't create issue/node/evidence records based on their selected mappings.

### Proposed solution
When the user submits the form, we process the import in a background job and then also show a console so that they can know the progress of the import.

Also, if a user submits the form without selecting an identifier, we tell them in the console.

### CODE REVIEW

File: app/views/dradis/plugins/csv/upload/new.html.erb [hunk 1]
```diff
@@ -65,3 +68,17 @@
...
+    <div id="status"></div>
+    <%= content_tag :div, id: 'result', style: 'display:none', data: { url: main_app.status_console_index_path }  do %>
+      <div id="console" class="mx-0 mb-0"></div>
+    <% end %>
+  </div>
+</div>
+
+<script type="text/javascript" charset="utf-8">
+  ConsoleUpdater.jobId = <%= @last_job %>;
+</script>
```
Comment (sean-yeoh) [inline app/views/dradis/plugins/csv/upload/new.html.erb:84]: I know it's weird, but I got it from: https://github.com/dradis/dradis-ce/blob/develop/app/views/upload/index.html.erb#L81-L83

File: app/controllers/dradis/plugins/csv/upload_controller.rb [hunk 2]
```diff
@@ -2,24 +2,44 @@ module Dradis::Plugins::CSV
...
 
-    before_action :load_attachment, only: [:new]
+    before_action :load_attachment, only: [:new, :create]
 
     def new
       @default_columns = ['Unique Identifier', 'Column Header From File', 'Type', 'Field in Dradis']
 
       @headers = ::CSV.open(@attachment.fullpath, &:readline)
+
+      @last_job = Log.new.uid
```
Comment (sean-yeoh) [inline app/controllers/dradis/plugins/csv/upload_controller.rb:12]: See: https://github.com/dradis/dradis-ce/blob/develop/app/controllers/upload_controller.rb#L20

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        next if rtp_import? && mapping['field'].blank?
+
+        field_name = rtp_import? ? mapping['field'] : row.headers[index.to_i]
+        field_value = row[index.to_i]
+        "#[#{field_name}]#\n#{field_value}"
+      end.compact.join("\n\n")
+    end
+
+    def rtp_import?
+      @mappings.all? { |index, mapping| mapping.keys.include?('field') }
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:79]: Would checking the project rtp work here instead? If not, this needs to be memoized.

File: app/controllers/dradis/plugins/csv/upload_controller.rb (hunk 2 above)
Comment (aapomm) [reply app/controllers/dradis/plugins/csv/upload_controller.rb:12]: This will fail for concurrent calls when a different user starts a different background job. Is this different from `params[:job_id]`?

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      @logger = Log.new(uid: uid)
+      @logger.write { "Job id is #{job_id}." }
+      @logger.write { 'Worker process starting background task.' }
+
+      @file = file
+      @identifier_col_index = identifier_col_index
+      @mappings = mappings
+      @project = Project.find(project_id)
+      node_col_mapping = @mappings.find { |index, field| field['type'] == 'node' }
+      @node_col_index = node_col_mapping[0].to_i if node_col_mapping
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:17]: Why not just call `first` after the the `find` instead of doing this in 2 steps?

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
+require 'csv'
+
+module Dradis::Plugins::CSV
+  class MappingImportJob < ApplicationJob
+    queue_as :dradis_project
+
+    def perform(file:, identifier_col_index:, mappings:, project_id:, uid:)
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:7]: Can you add in a comment here what is the expected mappings format?

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    def content_service
+      @content_service ||= Dradis::Plugins::ContentService::Base.new(
+        project: @project,
+        plugin: Dradis::Plugins::CSV
+      )
+    end
+
+    def import_csv!
+      @issue_mappings = @mappings.select { |index, mapping| mapping['type'] == 'issue' }
+      @evidence_mappings = @mappings.select { |index, mapping| mapping['type'] == 'evidence' }
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:40]: Alpha and I think we can drop plural here since each is now a single mapping.

```suggestion
      @evidence_mapping = @mappings.select { |index, mapping| mapping['type'] == 'evidence' }
      @issue_mapping = @mappings.select { |index, mapping| mapping['type'] == 'issue' }
```

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        @logger.info{ "\t => Creating new evidence (plugin_id: #{identifier})" }
+        @logger.info { "\t\t => Issue: #{issue.title} (issue_id: #{issue.to_issue.id})" }
+        @logger.info { "\t\t => Node: #{node.label} (node_id: #{node.id})" }
+
+        evidence_content = build_text(mappings: @evidence_mappings, row: row)
+        content_service.create_evidence(issue: issue, node: node, content: evidence_content)
+      end
+    end
+
+    def build_text(mappings:, row:)
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:68]:
```suggestion
    def build_text(mapping:, row:)
```

File: spec/jobs/dradis/plugins/csv/mapping_import_job_spec.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+          perform_job
+        }.to change(Issue, :count).by(0)
+
+        expect(Node.count).to eq(0)
+        expect(Evidence.count).to eq(0)
+      end
+    end
+  end
+end
+
```
Comment (aapomm) [inline spec/jobs/dradis/plugins/csv/mapping_import_job_spec.rb:105]:
```suggestion
```

File: app/controllers/dradis/plugins/csv/upload_controller.rb [hunk 9]
```diff
@@ -2,24 +2,44 @@ module Dradis::Plugins::CSV
...
     end
 
     def create
-      redirect_to main_app.project_upload_manager_path(current_project)
+      job_logger.write 'Enqueueing job to start in the background.'
+
+      MappingImportJob.perform_later(
+        file: @attachment.fullpath.to_s,
+        identifier_col_index: mappings_params[:identifier],
```
Comment (aapomm) [inline app/controllers/dradis/plugins/csv/upload_controller.rb:20]: Too wordy for a param. What about `:identifier`, `:id_index`, `:id_col`? `id` is already used as the parameter for the content service so I think we can use that.

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  class MappingImportJob < ApplicationJob
+    queue_as :dradis_project
+
+    def perform(file:, identifier_col_index:, mappings:, project_id:, uid:)
+      @logger = Log.new(uid: uid)
+      @logger.write { "Job id is #{job_id}." }
+      @logger.write { 'Worker process starting background task.' }
+
+      @file = file
+      @identifier_col_index = identifier_col_index
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:13]:
```suggestion
      @identifier_col_index = identifier_col_index.to_i
```
So in the class, we don't have to typecast it anymore.

File: app/controllers/dradis/plugins/csv/upload_controller.rb (hunk 2 above)
Comment (sean-yeoh) [reply app/controllers/dradis/plugins/csv/upload_controller.rb:12]: Yes, it's different. It's how `/uploads` work console work. See: https://github.com/dradis/dradis-csv-importer/pull/2#discussion_r925309351

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
+require 'csv'
```
[the remark this answers is not in the document]
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:1]: Removed.

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 4 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:17]: Good idea, thanks

Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:17]: Actually, this doesn't work because `nil.to_i` evaluates to `0` and it'll use the first column as the node label.

Would this be better?
```ruby
      @node_index = 
        if node_mapping = @mappings.find { |index, field| field['type'] == 'node' }&.first
          node_mapping[0].to_i
        end
```

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 3 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:79]: Thanks, that's much better

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 5 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:7]: Done

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 6 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:40]: It's not a single mapping. 

Mappings has the follow format:
```ruby
@mappings = {
  '0' => { 'type' => 'node' },
  '1' => { 'type' => 'issue', field: 'Title' },
  '2' => { 'type' => 'issue', field: 'Description' },
  '3' => { 'type' => 'evidence', field: 'Port' },
  '4' => { 'type' => 'evidence', field: 'Location' }
}
```

`@mappings.select { |index, mapping| mapping['type'] == 'issue' }` will return only issue mappings: e.g.
```ruby
{
  '1' => { 'type' => 'issue', field: 'Title' },
  '2' => { 'type' => 'issue', field: 'Description' }
}
```

Comment (aapomm) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:40]: I was thinking that this is a set of mappings:
```
@mappings = {
  '0' => { 'type' => 'node' },
  '1' => { 'type' => 'issue', field: 'Title' },
  '3' => { 'type' => 'evidence', field: 'Port' },
}
```

This is a mapping (singular) of issue fields:
```
{
  '1' => { 'type' => 'issue', field: 'Title' },
  '2' => { 'type' => 'issue', field: 'Description' }
}
```

This is a mapping (singular) of evidence fields:
```
{
  '3' => { 'type' => 'evidence', field: 'Port' },
  '4' => { 'type' => 'evidence', field: 'Location' }
}
```

This is a very minor thing so not a hill I'm willing to die on 😀

File: app/controllers/dradis/plugins/csv/upload_controller.rb (hunk 2 above)
Comment (aapomm) [reply app/controllers/dradis/plugins/csv/upload_controller.rb:12]: This makes more sense after re-reading. `@last_job` refers to a new job and not the last job that was completed. If so, can we use a different name instead of following the upload_controller naming?

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 6 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:40]: Ah okay, I think i'm just gonna leave it because i'm already using the `mapping` var (singular) inside the blocks already.

File: app/controllers/dradis/plugins/csv/upload_controller.rb (hunk 2 above)
Comment (sean-yeoh) [reply app/controllers/dradis/plugins/csv/upload_controller.rb]: Gotcha, i've updated it to `last_log` instead.

Comment (aapomm) [reply app/controllers/dradis/plugins/csv/upload_controller.rb]: Don't use `last`. What about `new_log_id` or `log_id` or `new_log`

File: app/controllers/dradis/plugins/csv/upload_controller.rb [hunk 12]
```diff
@@ -2,24 +2,51 @@ module Dradis::Plugins::CSV
...
+        file: @attachment.fullpath.to_s,
+        id_index: mappings_params[:identifier],
+        mappings: mappings_params[:mappings].to_h,
+        project_id: current_project.id,
+        uid: params[:log_uid].to_i
+      )
+
+      Resque.redis.del(params[:job_id])
+
+      head :ok
```
Comment (aapomm) [inline app/controllers/dradis/plugins/csv/upload_controller.rb:28]: What about rendering a create.js.erb here like in the UploadController where it initiates the ConsoleUpdater? That way we free up some code in the ever-growing upload.js

File: app/controllers/dradis/plugins/csv/upload_controller.rb [hunk 13]
```diff
@@ -2,24 +2,51 @@ module Dradis::Plugins::CSV
...
 
-    before_action :load_attachment, only: [:new]
+    before_action :load_attachment, only: [:new, :create]
 
     def new
       @default_columns = ['Unique Identifier', 'Column Header From File', 'Type', 'Field in Dradis']
 
       @headers = ::CSV.open(@attachment.fullpath, &:readline)
+
+      @log_uid = Log.new.uid + 1
```
Comment (aapomm) [inline app/controllers/dradis/plugins/csv/upload_controller.rb:12]: Why are we adding a `+ 1` here when the Log initialization already does that for us?

File: app/assets/javascripts/dradis/plugins/csv/upload.js [hunk 14]
```diff
@@ -34,5 +34,57 @@ document.addEventListener('turbolinks:load', function() {
...
+
+        $('[data-behavior~=view-content]').animate({
+          scrollTop: $('[data-behavior~=mapping-validation-messages]').scrollTop()
+        });
+      }
+
+      return valid;
+    });
+
+    function validateUniqueId() {
```
Comment (aapomm) [inline app/assets/javascripts/dradis/plugins/csv/upload.js:62]: Do we really need this whole block of code instead of just adding `required: true` for the `f.radio_button` call?

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+    # mappings hash:
+    # The key is the column index, with a hash containing the type of resource (evidence/issue/node)
+    # its supposed to map to and the dradis field for the resource (only for evidence and issues).
+    #
+    # e.g.
+    # {
+    # '0' => { 'type' => 'node' },
+    # '1' => { 'type' => 'issue', field: 'Title' },
+    # '2' => { 'type' => 'evidence', field: 'Port' }
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:13]:
```suggestion
    # '1' => { 'type' => 'issue', 'field' => 'Title' },
    # '2' => { 'type' => 'evidence', 'field' => 'Port' }
```
or is the `field` key needed to be a symbol?

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 16]
```diff
@@ -0,0 +1,91 @@
+module Dradis::Plugins::CSV
+  class MappingImportJob < ApplicationJob
+    queue_as :dradis_project
+
+    # mappings hash:
+    # The key is the column index, with a hash containing the type of resource (evidence/issue/node)
+    # its supposed to map to and the dradis field for the resource (only for evidence and issues).
```
Comment (aapomm) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:7]:
```suggestion
    # The key is the column index, while the value is a hash containing the type of resource (evidence/issue/node).
    # It's used to map a CSV header to a field in Dradis (only for evidence and issues).
```

File: app/controllers/dradis/plugins/csv/upload_controller.rb (hunk 13 above)
Comment (sean-yeoh) [reply app/controllers/dradis/plugins/csv/upload_controller.rb:12]: Gotcha, will remove it. I thought it didn't because there's a `+ 1` in `uploads.js.coffee` https://github.com/secroots/dradis-pro/blob/develop/app/assets/javascripts/tylium/modules/uploads.js.coffee#L28 :man_shrugging:

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 15 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:13]: It's string only, will update the comment.

File: app/controllers/dradis/plugins/csv/upload_controller.rb (hunk 12 above)
Comment (sean-yeoh) [reply app/controllers/dradis/plugins/csv/upload_controller.rb]: Agree, moved to `create.js.erb`.

File: app/assets/javascripts/dradis/plugins/csv/upload.js (hunk 14 above)
Comment (sean-yeoh) [reply app/assets/javascripts/dradis/plugins/csv/upload.js]: :man_facepalming: I forgot about `required: true`.

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+      unless @id_index
+        @logger.fatal('Unique Identifier doesn\'t exist, please choose a column as the Unique Identifier.')
+        return
+      end
+
+      @id_index = @id_index.to_i
+      @file = file
+      @project = Project.find(project_id)
+
+      import_csv!
```
Comment (etdsoft) [inline app/jobs/dradis/plugins/csv/mapping_import_job.rb:38]: isn't this what the engine's Importer should do?

The smell was having to define a ContentService instance inside a Job. That's a task solved by `dradis-plugins` and `Dradis::Plugins::Upload::Importer`

File: app/controllers/dradis/plugins/csv/upload_controller.rb [hunk 18]
```diff
@@ -2,27 +2,46 @@ module Dradis::Plugins::CSV
   class UploadController < ::AuthenticatedController
     include ProjectScoped
 
-    before_action :load_attachment, only: [:new]
+    before_action :load_attachment, only: [:new, :create]
 
     def new
       @default_columns = ['Column Header From File', 'Type', 'Field in Dradis']
 
       @headers = ::CSV.open(@attachment.fullpath, &:readline)
```
[the remark this answers is not in the document]
Comment (sean-yeoh) [reply app/controllers/dradis/plugins/csv/upload_controller.rb:10]: We could use `CSV.read(attachment, headers: true).headers` as `CSV.read calls `CSV.open` and passes a block to it.

[the remark this answers is not in the document]
Comment (sean-yeoh) [reply app/controllers/dradis/plugins/csv/upload_controller.rb:10]: Actually, the `&:readline` acts as a block.

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+      # Converts mapping hash into groups of arrays
+      # {
+      #   'node' => [['0', { 'type' => 'node' }]],
+      #   'identifier' => [['2', { 'type' => 'identifier' }]]
+      # }
+      @mappings_groups = mappings.group_by { |index, mapping| mapping['type'] }
+      @id_index = Integer(@mappings_groups['identifier']&.first&.first, exception: false)
+
+      unless @id_index
```
[the remark this answers is not in the document]
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb:29]: Is there a way to skip [dradis/pro/rules/content_service_callbacks.rb](https://github.com/secroots/dradis-pro/blob/develop/engines/dradispro-rules/lib/dradis/pro/rules/content_service_callbacks.rb)?

Creating an issue without `id` will cause it to raise an error from https://github.com/secroots/dradis-pro/blob/develop/engines/dradispro-rules/app/models/dradis/pro/rules/finding.rb#L98:

image

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    # }
+    def perform(file:, mappings:, project_id:, uid:)
+      @logger = Log.new(uid: uid)
+      @logger.write { "Job id is #{job_id}." }
+      @logger.write { 'Worker process starting background task.' }
+
+      # Converts mapping hash into groups of arrays
+      # {
+      #   'node' => [['0', { 'type' => 'node' }]],
+      #   'identifier' => [['2', { 'type' => 'identifier' }]]
```
[the remark this answers is not in the document]
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb]: Yes, there can only be 1 `node` and 1 `identifier`, which comes from the form.

File: app/jobs/dradis/plugins/csv/mapping_import_job.rb (hunk 17 above)
Comment (sean-yeoh) [reply app/jobs/dradis/plugins/csv/mapping_import_job.rb]: Moved to the `Importer` class.
